### ISSUE TITLE
Setup Result Proxy - Help

### ISSUE THREAD

**Issue by mrtargaryen** [searx/searx#991]
<img width="534" alt="screen shot 2017-08-12 at 12 20 54 pm" src="https://user-images.githubusercontent.com/23545795/29237027-dbea51cc-7f58-11e7-91ea-f5aca4c6c6cb.png">

Morty is installed on my Mac.

My Macs local IP is 192.168.1.128

I installed Docker for macOS, Kitematic, and VirtualBox  and created an instance of Searx. The local Address for Searx is http://192.168.99.100:32769/ ... I am able to use Searx, its working well.

However, I am attempting to setup Result Proxy for all search results so I can safely view non https sites.

I was given this link to follow in order to do this:

https://asciimoo.github.io/searx/admin/morty.html#how-to-setup-result-proxy

In searx’s settings.yml I have un-commented (enabled) the following:

result_proxy:
    url : http://127.0.0.1:3000/
    key : your_morty_proxy_key

Doing this did not enable the results search proxy.

Now I am not sure what I am suppose to put where it says "your_morty_proxy_key". Do I leave it as it is written or is there a morty proxy key to be found somewhere and I input that in "key : your_morty_proxy_key"

If that is the case where do I find this proxy key? Theres no indication where to find it on this link https://asciimoo.github.io/searx/admin/morty.html#how-to-setup-result-proxy

Again thank you for the time you've taken to read and help me figure this out.

**Comment by dalf** [comment]
The key is a secret shared between morty and searx.
The purpose is to avoid abuse of the morty proxy : only searx can create valid links to morty.

For example one link from searx.me :
https://searx.me:3000/?mortyurl=https%3A%2F%2Fasciimoo.github.io%2Fsearx%2F&mortyhash=e32499a7c4f6241f9dfa39cd577859bff01d137e13d9c493e9bb75d1b997d681

The mortyhash is created using the mortyurl and the shared key.
Morty on searx.me:3000 will check that the mortyhash matches the mortyurl and the shared key. 

To sum up : 
* define a key (whatever you want) on settings.yml
* when you start morty add the  ```-key``` parameter and the shared key.

**Comment by mrtargaryen** [comment]
> To sum up :

    define a key (whatever you want) on settings.yml
    when you start morty add the -key parameter and the shared key.

So lets say I make my key "key123", in the settings.yml I place http://127.0.0.1:3000/ (or do I place my searx url:3000?) and add my new key "key123"

And what do you mean by when I start morty add the -key parameter and shared key, where in the settings of morty do I do that?

Sorry, I understand this might be frustrating for you but I'm struggling to understand.

**Comment by asciimoo** [comment]
> So lets say I make my key "key123", in the settings.yml I place http://127.0.0.1:3000/ (or do I place my searx url:3000?) and add my new key "key123"

In this case your settings.yml have to contain these lines:
```
result_proxy:
    url : http://127.0.0.1:3000/
    key : key123
```
and execute the following command (in shell) to start morty:
```
"$GOPATH/bin/morty" -key "key123" -listen "127.0.0.1:3000"
```

**Comment by mrtargaryen** [comment]
> and use the following command to start morty:

"$GOPATH/bin/morty" -key "key123" -listen "127.0.0.1:3000"

Ahhh thank you so much, now I understand.

**Comment by mrtargaryen** [comment]
I edited the settings.yml 
<img width="750" alt="settings yml" src="https://user-images.githubusercontent.com/23545795/29246513-6ba1097c-8040-11e7-9fcb-50b18ac6a892.png">


Then I activated morty as per your instructions 
<img width="562" alt="activating morty" src="https://user-images.githubusercontent.com/23545795/29246516-853994c6-8040-11e7-95ba-2087a13f0fdd.png">

I restarted Searx

The I tested Searx to see if the proxy was working
<img width="1367" alt="searx results" src="https://user-images.githubusercontent.com/23545795/29246522-a42acc42-8040-11e7-9fa2-5156c5e043ed.png">



I'm not sure what I am doing wrong, I followed what you wrote.

**Comment by asciimoo** [comment]
I think there are at least two errors. First, you have a space before the config entry `result_proxy` in settings.yml, which causes config error for me if I restart searx with it. So, probably the second error is that you edit a wrong file or you don't restart searx after settings has been changed.

**Comment by mrtargaryen** [comment]
When I typed in http://127.0.0.1:3000/ in to the url bar in firefox I got this message.




![firefox_screenshot_2017-08-14t08-00-46 281z](https://user-images.githubusercontent.com/23545795/29263053-9671141e-811a-11e7-86b2-24f0fec8ba91.png)

**Comment by dalf** [comment]
That's normal : morty expects a mortyhash parameter (given by searx).

It seems you built your own docker image. Have you built it again after you modified the settings.yml ?

**Comment by mrtargaryen** [comment]
> It seems you built your own docker image. Have you built it again after you modified the settings.yml ?

I uninstalled Docker and morty and then reinstalled everything following this guide:

https://github.com/asciimoo/searx/wiki/Installation

I then modified the settings.yml, but I did not then build it again.

**Comment by porcupinepete** [comment]
Maybe I should just uninstall it all and try once more.

**Comment by mrtargaryen** [comment]
Ok,

So, I followed this installation page https://github.com/asciimoo/searx/wiki/Installation


Firstly I installed Docker: https://store.docker.com/editions/community/docker-ce-desktop-mac


Then I installed GO: https://golang.org/doc/install


Then in Terminal I ran this command: 

`git clone https://github.com/asciimoo/searx.git`


After that I ran this command:

`cd searx`


Then I ran this command:

`docker build -t searx` 


After this I ran this command to start SearX:


`docker run -d -p 8888:8888 searx`


Then I installed Morty: https://github.com/asciimoo/morty


Morty Installation and setup:

`$ go get github.com/asciimoo/morty`

`$ "$GOPATH/bin/morty" --help`


Test:

`$ cd "$GOPATH/src/github.com/asciimoo/morty"`

`$ go test`


Benchmark:

`$ cd "$GOPATH/src/github.com/asciimoo/morty"`

`$ go test -benchmem -bench .`


Then I edited the settings.yml file in Searx (https://asciimoo.github.io/searx/admin/morty.html#how-to-setup-result-proxy)

```
result_proxy:
    url : http://127.0.0.1:3000/
    key : 7720517
```

Then I executed this command in Terminal:

`"$GOPATH/bin/morty" -key "7720517" -listen "127.0.0.1:3000"`


I installed Searx on my iMac 27" late 2015 macOS 10.12.6. and I also tested this on my Ubuntu 16.04 PC and in both cases the results search proxy does not work.

These are exactly the stepss I took to get the Result Search Proxy to work in Searx. It does not work.

**Comment by mrtargaryen** [comment]
After running `/Users/Rick/go/bin/morty -key 7720517 -listen 127.0.0.1:3000` in Terminal (and then restarting Searx (`$ docker stop searx` and `$ docker start searx`) I accessed `/Users/Rick/go/bin/morty` and this is what I found;

```
Last login: Fri Aug 18 03:22:54 on ttys001
Rick-iMac:~ imac$ /Users/rick/go/bin/morty ; exit;
2017/08/18 05:16:20 listening on 127.0.0.1:3000
2017/08/18 05:16:20 Error in ListenAndServe:listen tcp4 127.0.0.1:3000: bind: address already in use
logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
Deleting expired sessions...none found.

[Process completed]
```

I'm not sure if its relevant but there appears to be an error.

# **Edit:**

Morty is using 127.0.0.1:3000 

I ran a Port Scan:

`Open TCP Port: 	3000   hbci`

I have set the Morty Key in settings.yml: 

<img width="818" alt="screen shot 2017-08-18 at 7 10 20 pm" src="https://user-images.githubusercontent.com/23545795/29452279-1b240410-8449-11e7-8fec-60e4bf451783.png">

and ran this in Terminal:

`/Users/Rick/go/bin/morty -key 7720517 -listen 127.0.0.1:3000`

(And I have stopped and started Searx).

So why isn't Morty receiving a mortyhash parameter from Searx? @dalf @asciimoo 

Can anyone replicate this issue? I have on a PC running Ubuntu.

**Comment by Pofilo** [comment]
A new docker repo can help you: https://github.com/searx/searx-docker.
Feel free to reopen the issue if needed.

**Comment by hubertwang** [comment]
Sorry to jump in..I have a question regarding the morty service.
If the service is correctly configured, the searx search result will have url for morty? Or need to manually calculate the mortyhash for the url?

**Comment by unixfox** [comment]
> Sorry to jump in..I have a question regarding the morty service.
> If the service is correctly configured, the searx search result will have url for morty? Or need to manually calculate the mortyhash for the url?

Please open a discussion for that: https://github.com/searx/searx/discussions

Don't revive a 4 years old issue just for that please.
